Update ViewHelper.js

alitan opened the pull request:
I tried to embed the editor's ViewHelper and the mouse events did not translate properly.

Instead of using a complex code to calc the mouse.x and mouse.y, i did use event.offsetX and event.offsetY This solved my problem and fixed any embedding issues.

alitan committed:
Update ViewHelper.js

I tried to embed the editor's ViewHelper and the mouse events did not translate properly.

Instead of using a complex code to calc the mouse.x and mouse.y, i did use event.offsetX and event.offsetY
This solved my problem and fixed any embedding issues.

mrdoob commented:
I've never had good results with `offsetX/offsetY`...

Are you sure that this new code isn't only fixing your use case but breaking all others?

mrdoob commented:
The editor in your `dev` branch seems to be broken:

https://raw.githack.com/alitan/three.js/dev/editor/index.html

If you click in the Y the camera goes to Z:

| | |
|---|---|
| ![Screenshot 2023-05-17 at 11 55 15 AM](https://github.com/mrdoob/three.js/assets/97088/340d3b06-c63f-4c4f-8196-8fc98d0c24c7) | ![Screenshot 2023-05-17 at 11 55 25 AM](https://github.com/mrdoob/three.js/assets/97088/b82321a8-ee8c-4331-988c-e7e4af114c85) |

In  if you click the Y the camera goes to Y:

https://raw.githack.com/mrdoob/three.js/dev/editor/index.html

| | |
|---|---|
| ![Screenshot 2023-05-17 at 11 55 43 AM](https://github.com/mrdoob/three.js/assets/97088/08f5cd00-803a-4524-8bbf-075098a655a7) | ![Screenshot 2023-05-17 at 11 55 54 AM](https://github.com/mrdoob/three.js/assets/97088/51aad14b-559e-491f-9fe0-1921d4cff9c5) |

Does that look right to you?

alitan committed:
Update ViewHelper.js

Y axis should be inversed.

alitan commented:
Sorry, forgot to inverse the Y axis.
I tested in many different embedding cases, OffsetX/OffsetY works better IMHO. (Also, the current logic is broken in certain DOM combinations)

mrdoob commented:
> Also, the current logic is broken in certain DOM combinations

Which certain DOM combination? Do you mind elaborating?

alitan commented:
When i put my custom editor (containing ViewHelper.js) inside nested divs (with flex layout and absolute positioning) the mouseX and mouseY is getting calculate wrongly (the positions getting shrink not coincide with underlying circles).

mrdoob commented:
I don't know... In my experience using `getBoundingClientRect()` tends to be more robust.

@Mugen87 I think I'd let you decide on this one.

alitan commented:
Is there any another way rather than using an invisible div over to capture the mouse events?
I am not sure if that is an elegant way IMHO compare to the rest of the (super minimalistic and elegant) code.

Mugen87 commented:
TBH, I always use `getBoundingClientRect()` and it works for me. Granted, there might be use cases where an alternative is required but I would not risk breakage in other apps by relying on `offsetX` and `offsetY`.

> Is there any another way rather than using an invisible div over to capture the mouse events?

Maybe a fiddle would help to understand what you are currently doing in your app. In this case, it's easier to find a different approach that works with the current version of `ViewHelper`.

alitan commented:
Ok guys, i found the bug.
There is a logic issue. it has nothing to do with mouse event data and/or getBoundingClientRect().
I realized that the mouse.y is getting calculated right with the current logic but mouse.x is the issue.

>> mouse.x = ( ( event.clientX - offsetX ) / ( rect.width - offsetX ) ) * 2 - 1;
      mouse.y = - ( ( event.clientY - offsetY ) / ( rect.bottom - offsetY ) ) * 2 + 1;

it should be (rect.right)

>> mouse.x = ( ( event.clientX - offsetX ) / ( rect.right - offsetX ) ) * 2 - 1;
      mouse.y = - ( ( event.clientY - offsetY ) / ( rect.bottom - offsetY ) ) * 2 + 1;

The reason that you guys didnt find this as a bug because the editor always running in full window width in example code.
The minute you put it in a container, rect.right is not the same as rect.width.

i will do a new commit.

alitan committed:
Update ViewHelper.js

Reverting back to the original code with one fix. rect.width should be rect.right

Mugen87 commented on examples/jsm/helpers/ViewHelper.js:
```diff
@@ -163,11 +163,10 @@ class ViewHelper extends Object3D {
 
 			if ( this.animating === true ) return false;
-
-			const rect = domElement.getBoundingClientRect();
```
Since you have deleted this line, the following runtime error happens now:
> ViewHelper.js:167 Uncaught ReferenceError: rect is not defined

Mugen87 replied:
In any event, using `rect.right` sounds like a good fix.

mrdoob committed:
Update ViewHelper.js

Mugen87 committed:
Update ViewHelper.js
